### ISSUE TITLE
padrange optimization makes incorrect opcode assumptions

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#12549]
Migrated from [rt.perl.org#115688](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=115688) (status was 'resolved')

Searchable as RT115688$

**Comment by p5pRT** [comment]
### From @mauke

#### Created by @mauke



This may be the same bug as \#115684\, which is \(nominally\) about Lexical&#8203;::Types\.
I'm reporting this separately because I think I understand what's breaking
Function&#8203;::Parameters\, at least\.

Testcase&#8203;:
\#\!perl
use warnings;
use strict;
use Function&#8203;::Parameters;
use Test&#8203;::More tests => 3;

my $outer = "outer";
fun foo\($bar\, $baz = $outer\) \{
&nbsp;    return "$bar $baz";
\}

is $outer\, "outer";
is foo\("A"\)\, "A outer";
is $outer\, "outer";
\_\_END\_\_

Running this with This is perl 5\, version 17\, subversion 6 \(v5\.17\.6
\(v5\.17\.5\-211\-gdf7075a\)\) built for i686\-linux built with debugging enabled\, I
get&#8203;:

perl&#8203;: op\.c&#8203;:11028&#8203;: Perl\_rpeep&#8203;: Assertion \`p\->op\_targ == base \+ count' failed\.

Without debugging the assert is compiled out and I get&#8203;:
1\.\.3
ok 1
Use of uninitialized value $baz in concatenation \(\.\) or string at wtf line 9\.
not ok 2
\# Failed test at wtf line 13\.
\# got&#8203;: 'A '
\# expected&#8203;: 'A outer'
not ok 3
\# Failed test at wtf line 14\.
\# got&#8203;: undef
\# expected&#8203;: 'outer'
\# Looks like you failed 2 tests of 3\.


The code in question from op\.c \(around line 11028\)&#8203;:

&nbsp;                else \{
&nbsp;                    if \(\(p\->op\_private & OPpLVAL\_INTRO\) \!= intro\)
&nbsp;                        break;
&nbsp;                    /\* we expect targs to be contiguous in my\($a\,$b\,$c\)
&nbsp;                     \* but not in \($a\, $x\, $z\)\. In the latter case\, stop
&nbsp;                     \* on the first non\-contiguous padop \*/
&nbsp;                    if \(\!intro && p\->op\_targ \!= base \+ count\)
&nbsp;                        break;
&nbsp;                    assert\(p\->op\_targ == base \+ count\);
&nbsp;                    /\* all the padops should be in the same context \*/
&nbsp;                    if \(gimme \!= \(p\->op\_flags & OPf\_WANT\)\)
&nbsp;                        break;
&nbsp;                \}

\.\.\. "we expect targs to be contiguous" \(hah\!\)

Well\, the targs aren't contiguous in this case and I think the "\!intro &&" part
should simply be removed \(along with the assert\)\. That is\, if p\->op\_targ \!=
base \+ count\, simply cancel the padrange optimization \(instead of aborting the
perl process \(debug mode\) or generating incorrect code \(non\-debug mode\)\)\.

I'd patch it myself\, but how do you write a test for this? I don't think this
can be triggered with the normal parser in perly\.y\, so maybe a custom keyword
in XS&#8203;::APItest? \(Also\, is my understanding of the issue correct?\)

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=high

Site configuration information for perl 5.17.6:

Configured by mauke at Mon Nov 12 01:05:35 CET 2012.

Summary of my perl5 (revision 5 version 17 subversion 6) configuration:
  Commit id: df7075a87da4c0a53dcce62713d9defecef39a46
  Platform:
    osname=linux, osvers=2.6.32-5-686, archname=i686-linux
    uname='linux debian-vb 2.6.32-5-686 #1 smp sun sep 23 09:49:36 utc 2012 i686 gnulinux '
    config_args=''
    hint=previous, useposix=true, d_sigaction=define
    useithreads=undef, usemultiplicity=undef
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=undef, use64bitall=undef, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O -g -DDEBUGGING',
    cppflags='-fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64'
    ccversion='', gccversion='4.7.1', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -fstack-protector -L/usr/local/lib'
    libpth=/usr/local/lib /lib/../lib /usr/lib/../lib /lib /usr/lib /usr/lib/i486-linux-gnu /usr/lib64
    libs=-lnsl -ldl -lm -lcrypt -lutil -lc
    perllibs=-lnsl -ldl -lm -lcrypt -lutil -lc
    libc=/lib/libc-2.11.3.so, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.11.3'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -L/usr/local/lib -fstack-protector'

Locally applied patches:
    


@INC for perl 5.17.6:
    /home/mauke/perl5/devel/df7075a87da4c0a53dcce62713d9defecef39a46/lib/site_perl/5.17.6/i686-linux
    /home/mauke/perl5/devel/df7075a87da4c0a53dcce62713d9defecef39a46/lib/site_perl/5.17.6
    /home/mauke/perl5/devel/df7075a87da4c0a53dcce62713d9defecef39a46/lib/5.17.6/i686-linux
    /home/mauke/perl5/devel/df7075a87da4c0a53dcce62713d9defecef39a46/lib/5.17.6
    .


Environment for perl 5.17.6:
    HOME=/home/mauke
    LANG=en_US.utf8
    LANGUAGE (unset)
    LC_COLLATE=POSIX
    LD_LIBRARY_PATH=/home/mauke/usr/lib
    LOGDIR (unset)
    PATH=/home/mauke/perl5/perlbrew/bin:/home/mauke/.cabal/bin:/home/mauke/usr/bin:/usr/local/bin:/usr/bin:/bin:/usr/local/games:/usr/games
    PERLBREW_BASHRC_VERSION=0.55
    PERLBREW_HOME=/home/mauke/.perlbrew
    PERLBREW_ROOT=/home/mauke/perl5/perlbrew
    PERL_BADLANG (unset)
    PERL_UNICODE=SAL
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @mauke

I've found a way to trigger this bug from userspace without additional
modules\. I've attached a patch that merely adds a failing test because
my first attempt at fixing the bug was unsuccessful\.

**Comment by p5pRT** [comment]
### From @mauke

<details><summary>0001-ensure-padrange-doesn-t-break-non-contiguous-targs.patch</summary>

```diff
From 43a48d798e26779b2baa4e6573e0095c3cadf133 Mon Sep 17 00:00:00 2001
From: Lukas Mai <l.mai@web.de>
Date: Mon, 12 Nov 2012 11:14:02 +0100
Subject: [PATCH] ensure padrange doesn't break non-contiguous targs

Turns out you can get non-contiguous targs even in pure Perl with code
like:

  my $foo;
  sub {
      my ($bar, 0 ? $foo : $baz);
  }

This commit adds a new test to make sure things work as before the
padrange optimization. (See also bug #115688).
---
 t/op/padrange.t |   20 ++++++++++++++++++++
 1 files changed, 20 insertions(+), 0 deletions(-)
 create mode 100644 t/op/padrange.t

diff --git a/t/op/padrange.t b/t/op/padrange.t
new file mode 100644
index 0000000..e530fa0
--- /dev/null
+++ b/t/op/padrange.t
@@ -0,0 +1,20 @@
+#!./perl -w
+
+BEGIN {
+    chdir 't';
+    @INC = '../lib';
+    require './test.pl';
+}
+use strict;
+
+BEGIN { plan tests => 3; }
+
+my $outer = 'outer';
+sub foo {
+    my ($bar, 0 ? $outer : $baz) = @_;
+    "$bar $outer $baz"
+}
+
+is $outer, 'outer';
+is foo('A', 'B'), 'A outer B';
+is $outer, 'outer';
-- 
1.7.2.5


```
</details>

**Comment by p5pRT** [comment]
@mauke - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @iabyn

On Sun\, Nov 11\, 2012 at 05&#8203;:06&#8203;:58PM \-0800\, l\.mai@&#8203;web\.de wrote&#8203;:
> \.\.\. "we expect targs to be contiguous" \(hah\!\)
> 
> Well\, the targs aren't contiguous in this case and I think the "\!intro &&" part
> should simply be removed \(along with the assert\)\. That is\, if p\->op\_targ \!=
> base \+ count\, simply cancel the padrange optimization \(instead of aborting the
> perl process \(debug mode\) or generating incorrect code \(non\-debug mode\)\)\.
> 
> I'd patch it myself\, but how do you write a test for this? I don't think this
> can be triggered with the normal parser in perly\.y\, so maybe a custom keyword
> in XS&#8203;::APItest? \(Also\, is my understanding of the issue correct?\)

Your understanding is correct\. I'll change the code\, but think I'll skip
adding a test \(just add a comment in the code\)\.

\-\- 
Gravity is just a theory; teach Intelligent Falling in our schools\!
&nbsp;    http&#8203;://www\.theonion\.com/content/node/39512

**Comment by p5pRT** [comment]
### From @iabyn

On Mon\, Nov 12\, 2012 at 10&#8203;:30&#8203;:30AM \+0000\, Dave Mitchell wrote&#8203;:
> On Sun\, Nov 11\, 2012 at 05&#8203;:06&#8203;:58PM \-0800\, l\.mai@&#8203;web\.de wrote&#8203;:
> > \.\.\. "we expect targs to be contiguous" \(hah\!\)
> > 
> > Well\, the targs aren't contiguous in this case and I think the "\!intro &&" part
> > should simply be removed \(along with the assert\)\. That is\, if p\->op\_targ \!=
> > base \+ count\, simply cancel the padrange optimization \(instead of aborting the
> > perl process \(debug mode\) or generating incorrect code \(non\-debug mode\)\)\.
> > 
> > I'd patch it myself\, but how do you write a test for this? I don't think this
> > can be triggered with the normal parser in perly\.y\, so maybe a custom keyword
> > in XS&#8203;::APItest? \(Also\, is my understanding of the issue correct?\)
> 
> Your understanding is correct\. I'll change the code\, but think I'll skip
> adding a test \(just add a comment in the code\)\.

Now done with 18c931a3833eccac01983e3e50239ca36de82ec4

\-\- 
O Unicef Clearasil\!
Gibberish and Drivel\!
&nbsp;    \-\- "Bored of the Rings"

**Comment by p5pRT** [comment]
### From @mauke

On 12\.11\.2012 12&#8203;:10\, Dave Mitchell wrote&#8203;:
> On Mon\, Nov 12\, 2012 at 10&#8203;:30&#8203;:30AM \+0000\, Dave Mitchell wrote&#8203;:
>> On Sun\, Nov 11\, 2012 at 05&#8203;:06&#8203;:58PM \-0800\, l\.mai@&#8203;web\.de wrote&#8203;:
>>> \.\.\. "we expect targs to be contiguous" \(hah\!\)
>>>
>>> Well\, the targs aren't contiguous in this case and I think the "\!intro &&" part
>>> should simply be removed \(along with the assert\)\. That is\, if p\->op\_targ \!=
>>> base \+ count\, simply cancel the padrange optimization \(instead of aborting the
>>> perl process \(debug mode\) or generating incorrect code \(non\-debug mode\)\)\.
>>>
>>> I'd patch it myself\, but how do you write a test for this? I don't think this
>>> can be triggered with the normal parser in perly\.y\, so maybe a custom keyword
>>> in XS&#8203;::APItest? \(Also\, is my understanding of the issue correct?\)
>>
>> Your understanding is correct\. I'll change the code\, but think I'll skip
>> adding a test \(just add a comment in the code\)\.
>
> Now done with 18c931a3833eccac01983e3e50239ca36de82ec4
>

I just ran a 'make coretest' and it's still failing my new op/padrange\.t 
test&#8203;:

1\.\.3
ok 1
Use of uninitialized value $outer in concatenation \(\.\) or string at 
t/op/padrange\.t line 15\.
not ok 2
\# Failed test 2 \- at t/op/padrange\.t line 19
\#      got "A  B"
\# expected "A outer B"
ok 3

Any idea what's going on?

**Comment by p5pRT** [comment]
### From @iabyn

On Mon\, Nov 12\, 2012 at 12&#8203;:46&#8203;:53PM \+0100\, Lukas Mai wrote&#8203;:
> On 12\.11\.2012 12&#8203;:10\, Dave Mitchell wrote&#8203;:
> >On Mon\, Nov 12\, 2012 at 10&#8203;:30&#8203;:30AM \+0000\, Dave Mitchell wrote&#8203;:
> >>On Sun\, Nov 11\, 2012 at 05&#8203;:06&#8203;:58PM \-0800\, l\.mai@&#8203;web\.de wrote&#8203;:
> >>>\.\.\. "we expect targs to be contiguous" \(hah\!\)
> >>>
> >>>Well\, the targs aren't contiguous in this case and I think the "\!intro &&" part
> >>>should simply be removed \(along with the assert\)\. That is\, if p\->op\_targ \!=
> >>>base \+ count\, simply cancel the padrange optimization \(instead of aborting the
> >>>perl process \(debug mode\) or generating incorrect code \(non\-debug mode\)\)\.
> >>>
> >>>I'd patch it myself\, but how do you write a test for this? I don't think this
> >>>can be triggered with the normal parser in perly\.y\, so maybe a custom keyword
> >>>in XS&#8203;::APItest? \(Also\, is my understanding of the issue correct?\)
> >>
> >>Your understanding is correct\. I'll change the code\, but think I'll skip
> >>adding a test \(just add a comment in the code\)\.
> >
> >Now done with 18c931a3833eccac01983e3e50239ca36de82ec4
> >
> 
> I just ran a 'make coretest' and it's still failing my new
> op/padrange\.t test&#8203;:
> 
> 1\.\.3
> ok 1
> Use of uninitialized value $outer in concatenation \(\.\) or string at
> t/op/padrange\.t line 15\.
> not ok 2
> \# Failed test 2 \- at t/op/padrange\.t line 19
> \#      got "A  B"
> \# expected "A outer B"
> ok 3
> 
> Any idea what's going on?

Note that your test fails on old pre\-padrange perls too\.

Your test code simplifies to&#8203;:

my $b='x';
\{
&nbsp;    my \($a\, 0 ? $b : $c\) = \(1\,2\,3\);
&nbsp;    print "a=$a b=$b c=$c\\n";
\}

and the central 'my' line is roughly equivalent to

&nbsp;    my \($a\, $b\, $c\);
&nbsp;    \($a\, $c\) = \(1\,2\,3\);

which gives expected result of a=1 b=\(undef\) c=2

Having said that\, my patch makes Concise crash\, so I think I'll need to
give it some more thought\.

\-\- 
"You may not work around any technical limitations in the software"
&nbsp;    \-\- Windows Vista license

**Comment by p5pRT** [comment]
### From @mauke

On 12\.11\.2012 14&#8203;:41\, Dave Mitchell wrote&#8203;:

> Note that your test fails on old pre\-padrange perls too\.

Argh\. You're right\, of course\. Sorry\, I must have been testing the wrong 
file and I withdraw the patch\.

But I've just realized your explanation of my \(0 ? $x : $y\) as my $x; my 
\($y\); isn't quite right&#8203;:

\#\!perl
use warnings;
use strict;

sub foo \{
&nbsp;         my \(0 ? $foo : $bar\) = @&#8203;\_;
&nbsp;         return $foo\+\+ \. '&#8203;:' \. $bar;
\}

print foo\($\_\)\, "\\n" for qw\(A B C\);
\_\_END\_\_
0&#8203;:A
1&#8203;:B
2&#8203;:C

This is another instance of the 'my $foo if 0' problem\, but written as 
'my \($foo if 0\)' here \(more or less\)\.

**Comment by p5pRT** [comment]
### From @cpansprout

On Mon Nov 12 08&#8203;:28&#8203;:54 2012\, mauke\- wrote&#8203;:
> On 12\.11\.2012 14&#8203;:41\, Dave Mitchell wrote&#8203;:
> 
> > Note that your test fails on old pre\-padrange perls too\.
> 
> Argh\. You're right\, of course\. Sorry\, I must have been testing the wrong 
> file and I withdraw the patch\.
> 
> But I've just realized your explanation of my \(0 ? $x : $y\) as my $x; my 
> \($y\); isn't quite right&#8203;:
> 
> \#\!perl
> use warnings;
> use strict;
> 
> sub foo \{
>          my \(0 ? $foo : $bar\) = @&#8203;\_;
>          return $foo\+\+ \. '&#8203;:' \. $bar;
> \}
> 
> print foo\($\_\)\, "\\n" for qw\(A B C\);
> \_\_END\_\_
> 0&#8203;:A
> 1&#8203;:B
> 2&#8203;:C
> 
> This is another instance of the 'my $foo if 0' problem\, but written as 
> 'my \($foo if 0\)' here \(more or less\)\.

I think my \(0 ? $x : $y\) should be a compilation error\.  Constant
folding shouldn’t change behaviour\.

That said\, this error message could be improved&#8203;:

$ \./perl \-ILib \-e 'my\($x?$y&#8203;:$z\)'
Can't declare null operation in "my" at \-e line 1\, at EOF
Execution of \-e aborted due to compilation errors\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @iabyn

On Mon\, Nov 12\, 2012 at 05&#8203;:28&#8203;:12PM \+0100\, Lukas Mai wrote&#8203;:
> But I've just realized your explanation of my \(0 ? $x : $y\) as my
> $x; my \($y\); isn't quite right&#8203;:

I never said that\. I said that

&nbsp;    my \(0 ? $x : $y\) = expr;

was \*roughly\* equivalent to 

&nbsp;    my \($x\, $y\);
&nbsp;    \($y\) = expr;

i\.e\. I was was trying to separate out the compile\-time effect of allocating
pad slots \(allocated for both $x\, $y\, masking any outer $x or $y\)\,
verses the runtime effect\, where only $y and not $x gets assigned to\. Of
course\, the run\-time effect of doing a SAVECLEARCV on $x gets skipped too\.

\-\- 
I've often wanted to drown my troubles\, but I can't get my wife to go
swimming\.

**Comment by p5pRT** [comment]
### From @jkeenan

On Tue Nov 13 04&#8203;:22&#8203;:11 2012\, davem wrote&#8203;:
> On Mon\, Nov 12\, 2012 at 05&#8203;:28&#8203;:12PM \+0100\, Lukas Mai wrote&#8203;:
> > But I've just realized your explanation of my \(0 ? $x : $y\) as my
> > $x; my \($y\); isn't quite right&#8203;:
> 
> I never said that\. I said that
> 
>     my \(0 ? $x : $y\) = expr;
> 
> was \*roughly\* equivalent to 
> 
>     my \($x\, $y\);
>     \($y\) = expr;
> 
> i\.e\. I was was trying to separate out the compile\-time effect of allocating
> pad slots \(allocated for both $x\, $y\, masking any outer $x or $y\)\,
> verses the runtime effect\, where only $y and not $x gets assigned to\. Of
> course\, the run\-time effect of doing a SAVECLEARCV on $x gets skipped too\.
> 


Dave\, Father C\, mauke&#8203;:  Could we get an update on the status of this ticket?

Thank you very much\.
Jim Keenan

**Comment by p5pRT** [comment]
### From @cpansprout

On Fri Dec 13 16&#8203;:38&#8203;:20 2013\, jkeenan wrote&#8203;:
> On Tue Nov 13 04&#8203;:22&#8203;:11 2012\, davem wrote&#8203;:
> > On Mon\, Nov 12\, 2012 at 05&#8203;:28&#8203;:12PM \+0100\, Lukas Mai wrote&#8203;:
> > > But I've just realized your explanation of my \(0 ? $x : $y\) as my
> > > $x; my \($y\); isn't quite right&#8203;:
> >
> > I never said that\. I said that
> >
> > my \(0 ? $x : $y\) = expr;
> >
> > was \*roughly\* equivalent to
> >
> > my \($x\, $y\);
> > \($y\) = expr;
> >
> > i\.e\. I was was trying to separate out the compile\-time effect of
> > allocating
> > pad slots \(allocated for both $x\, $y\, masking any outer $x or $y\)\,
> > verses the runtime effect\, where only $y and not $x gets assigned to\.
> > Of
> > course\, the run\-time effect of doing a SAVECLEARCV on $x gets skipped
> > too\.
> >
> 
> 
> Dave\, Father C\, mauke&#8203;:  Could we get an update on the status of this
> ticket?

It can be closed \(doing so now\)\.  The commit Dave cited fixed the issue at hand\.  My only involvement in this ticket was a side comment\, that I have now addressed in commit 4748e00\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'
